crates/wysiwyg/src/composer_model/hyperlinks.rs

```diff
@@ -59,15 +59,84 @@ where
         suggestion: SuggestionPattern,
         attributes: Vec<(S, S)>,
     ) -> ComposerUpdate<S> {
-        // TODO - this function allows us to accept a Vec of attributes to add to the Link we create,
-        // but these attributes will be present in the html of the message we output. We may need to
-        // add a step in the future that strips these attributes from the html before it is sent.
+        // This function removes the text between the suggestion start and end points, updates the cursor position
+        // and then calls set_mention_with_text (for equivalence with stages for inserting a link)
 
         self.do_replace_text_in(S::default(), suggestion.start, suggestion.end);
         self.state.start = Location::from(suggestion.start);
         self.state.end = self.state.start;
-        self.set_mention_with_text(url, text, attributes);
-        self.do_replace_text(" ".into())
+
+        self.set_mention_with_text(url, text, attributes)
+    }
+
+    fn set_mention_with_text(
+        &mut self,
+        url: S,
+        text: S,
+        attributes: Vec<(S, S)>,
+    ) -> ComposerUpdate<S> {
+        // this function is similar to set_link_with_text, but now we call a new simpler insertion method
+        self.push_state_to_history();
+        self.insert_node_at_cursor(DomNode::new_mention(url, text, attributes))
+    }
+
+    /// Inserts the given node at the current cursor position
+    fn insert_node_at_cursor(&mut self, node: DomNode<S>) -> ComposerUpdate<S> {
+        let (s, e) = self.safe_selection();
+        let range = self.state.dom.find_range(s, e);
+
+        // manually determine where the cursor will be after we finish the amendment
+        let new_cursor_position = s + node.text_len();
+
+        let mut should_add_trailing_space = false;
+
+        // do nothing if we don't have a cursor
+        // TODO determine behaviour if we try to insert when we have a selection
+        if range.is_selection() {
+            ComposerUpdate::<S>::keep();
+        }
+
+        // manipulate the state of the dom as required
+        if range.is_empty() {
+            // this happens if we replace the whole of a single text node in the composer
```

Comment:
I'm a bit confused about what this means. Wouldn't an empty range just imply there are no nodes in the DOM?